torch/distributed/_shard/checkpoint/metadata.py

```diff
@@ -0,0 +1,90 @@
+import io
+import pickle
+from dataclasses import dataclass
+from typing import Dict, List, Tuple, Union
+
+import torch
+from torch.distributed._shard.sharded_tensor import (
+    ShardedTensor,
+    ShardedTensorMetadata,
+    ShardMetadata,
+)
+
+TENSOR_TYPE = Union[torch.Tensor, ShardedTensor]
+
+@dataclass
+class ShardStorageMetadata:
+    shard_metadata: ShardMetadata
+    # storage key used for this particular Shard
+    storage_key: str
+    # Length in bytes for this shard
+    length: int
+
+
+# Metadata for each param.
+@dataclass
```

Comment:
A bit confused what this `ShardedTensorStorageMetadata` is supposed to handle.

Ostensibly, each storage implementation may want to have its own specific details included (e.g. a file system based may want to add file name mapping and an S3 based may want to know the bucket name).

It's not possible for the torch.dist version to handle everything. To me torch.distributed should only define the high level logical tensor metadata contract (dtype, strides, sizes etc. for a shard) and nothing beyond that. Storage layer should be free to choose whatever additional metadata to be "fused" in and it's not restricted to be 1-1 (a shard `foo` could be chunked into 5 files for example so the loading logic should know the mapping between that and the 5 files).

Reply:
This only features the metadata that's not specific of any storage implementation.

Reply:
Maybe help us understand why is `ShardMetadata` placed in `ShardStorageMetadata`?
I understand why we need maybe `storage_key` or `length` but `ShardedTensorMetadata` already has a list of `ShardMetadata`. What's the singleton in storage metadata for?

Reply:
The core issue here is that there's no guarantee on ordering or equivalency between the `ShardMetatada` list and the `ShardStorageMetadata` list.

We could replace that `ShardeMetadata` field with an index into the ShardedTensorMetadata.
It would complicate things a bit but avoid the redundant data.

Reply:
One option would be to not use ShardedTensorMetadata at all but some of its internal properties, so something like this to avoid duplication:

```
@dataclass
class ShardedTensorStorageMetadata(object):
    shards_metadata_storage: List[ShardStorageMetadata] = field(default_factory=list)
    size: torch.Size = field(default=torch.Size([]))
    tensor_properties: TensorProperties = field(
        default=TensorProperties(dtype=torch.get_default_dtype(),
                                 layout=torch.strided,
                                 requires_grad=False,
                                 memory_format=torch.contiguous_format,
                                 pin_memory=False))
```

We could have simple methods to convert from this to ShardedTensorMetadata and vice-versa.

Reply:
@pritamdamania87  that makes sense, specially once we switch to local planning.